simpleclient/src/main/java/io/prometheus/client/TimeWindowQuantiles.java

```diff
@@ -11,44 +14,56 @@
-  private int currentBucket;
-  private long lastRotateTimestampMillis;
+  private final ConcurrentLinkedQueue<CKMSQuantiles> buckets;
+  private final AtomicLong lastRotateTimestampMillis;
   private final long durationBetweenRotatesMillis;
 
   public TimeWindowQuantiles(Quantile[] quantiles, long maxAgeSeconds, int ageBuckets) {
     this.quantiles = quantiles;
-    this.ringBuffer = new CKMSQuantiles[ageBuckets];
+    this.buckets = new ConcurrentLinkedQueue<CKMSQuantiles>();
     for (int i = 0; i < ageBuckets; i++) {
-      this.ringBuffer[i] = new CKMSQuantiles(quantiles);
+      this.buckets.add(new CKMSQuantiles(quantiles));
     }
-    this.currentBucket = 0;
-    this.lastRotateTimestampMillis = System.currentTimeMillis();
+    this.lastRotateTimestampMillis = new AtomicLong(System.currentTimeMillis());
     this.durationBetweenRotatesMillis = TimeUnit.SECONDS.toMillis(maxAgeSeconds) / ageBuckets;
   }
 
   public double get(double q) {
-    CKMSQuantiles currentBucket = rotate();
-    return currentBucket.get(q);
+    // On concurrent `get` and `rotate`, it is acceptable to `get` the sample from an outdated `bucket`.
+    return getCurrentBucket().get(q);
   }
 
   public void insert(double value) {
+    // On concurrent `insert` and `rotate`, it might be acceptable to lose the measurement in the newest `bucket`.
     rotate();
-    for (CKMSQuantiles ckmsQuantiles : ringBuffer) {
-      ckmsQuantiles.insert(value);
+
+    for (CKMSQuantiles bucket : buckets) {
+      bucket.insert(value);
     }
   }
 
-  private synchronized CKMSQuantiles rotate() {
-    long timeSinceLastRotateMillis = System.currentTimeMillis() - lastRotateTimestampMillis;
-    while (timeSinceLastRotateMillis > durationBetweenRotatesMillis) {
-      ringBuffer[currentBucket] = new CKMSQuantiles(quantiles);
-      if (++currentBucket >= ringBuffer.length) {
-        currentBucket = 0;
+  private CKMSQuantiles getCurrentBucket() {
+    rotate();
+
+    return buckets.peek();
+  }
+
+  private void rotate() {
+    // On concurrent `rotate` and `rotate`:
+    //  - `currentTimeMillis` is cached to reduce thread contention.
+    //  - `lastRotateTimestampMillis` is used to ensure the correct number of rotations.
+    long currentTimeMillis = System.currentTimeMillis();
+    long lastRotateTimestampMillis = this.lastRotateTimestampMillis.get();
+    while (currentTimeMillis - lastRotateTimestampMillis > durationBetweenRotatesMillis) {
+      if (this.lastRotateTimestampMillis.compareAndSet(
+          lastRotateTimestampMillis, lastRotateTimestampMillis + durationBetweenRotatesMillis)) {
+        CKMSQuantiles bucket = new CKMSQuantiles(quantiles);
+        // rotate buckets (not atomic)
+        buckets.add(bucket);
+        buckets.remove();
       }
-      timeSinceLastRotateMillis -= durationBetweenRotatesMillis;
-      lastRotateTimestampMillis += durationBetweenRotatesMillis;
+      lastRotateTimestampMillis = this.lastRotateTimestampMillis.get();
```

Comment:
Wasn't this already set in the compareAndSet?

Reply:
I'm not too good at naming things, sorry for the confusion. :(

If `compareAndSet` succeeded, then `this.lastRotateTimestampMillis` was set.
If `compareAndSet` failed, then `this.lastRotateTimestampMillis` had a different value.
In both cases, local `lastRotateTimestampMillis` still has the old stale value.

It looks like `compareAndExchange` is Java 9, it returns the previous value to avoid the `get`. 
It doesn't looks like `compareAndSetAndGet` is a thing. 😃

Reply:
Just "last" would be clearer.